canary/mqtt5/canary.ts

```diff
@@ -13,207 +13,263 @@ type Args = { [index: string]: any };
+    totalOperation: number;
+}
+
+interface TestContext {
+    duration: number;
+    hostname: string;
+    port: number;
+    tps_sleep_time: number;
+    clients: number;
 }
 
 interface CanaryContext {
-    client : mqtt5.Mqtt5Client;
+    clients: mqtt5.Mqtt5Client[];
 
-    mqttStats : CanaryMqttStatistics;
+    mqttStats: CanaryMqttStatistics;
 
-    subscriptions: string[];
+    subscriptions: string[][];
 }
 
-function createCanaryClient(mqttStats : CanaryMqttStatistics) : mqtt5.Mqtt5Client {
-    const client_config : mqtt5.Mqtt5ClientConfig = {
-        hostName : process.env.AWS_TEST_MQTT5_DIRECT_MQTT_HOST ?? "localhost",
-        port : parseInt(process.env.AWS_TEST_MQTT5_DIRECT_MQTT_PORT ?? "0")
+function sleep(millisecond: number) {
+    return new Promise((resolve) => setInterval(resolve, millisecond));
+}
+
+function getRandomIndex(clients : mqtt5.Mqtt5Client[]): number
+{
+    return Math.floor(Math.random() * clients.length);
+}
+
+function createCanaryClients(testContext: TestContext, mqttStats: CanaryMqttStatistics): mqtt5.Mqtt5Client[] {
+    const client_config: mqtt5.Mqtt5ClientConfig = {
+        hostName: testContext.hostname,
+        port: testContext.port
     };
 
-    let client : mqtt5.Mqtt5Client = new mqtt5.Mqtt5Client(client_config);
+    const clients = [];
 
-    client.on('error', (error: ICrtError) => {});
-    client.on("messageReceived",(message: mqtt5_packet.PublishPacket) : void => {
-        mqttStats.publishesReceived++;
-    });
+    for (let i = 0; i < testContext.clients; i++) {
+        let client: mqtt5.Mqtt5Client = new mqtt5.Mqtt5Client(client_config);
+
+        client.on('error', (error: ICrtError) => { });
+        client.on("messageReceived", (eventData: mqtt5.MessageReceivedEvent): void => {
+            mqttStats.publishesReceived++;
+        });
+
+        ++mqttStats.clientsUsed;
 
-    return client;
+        clients.push(client);
+    }
+
+    return clients;
 }
 
-async function doSubscribe(context : CanaryContext) {
+async function doSubscribe(context: CanaryContext) {
     let topicFilter: string = `Mqtt5/Canary/RandomSubscribe${uuid()}`;
 
+    let index = getRandomIndex(context.clients);
     try {
         context.mqttStats.subscribesAttempted++;
 
-        await context.client.subscribe({
+        await context.clients[index].subscribe({
             subscriptions: [
-                {topicFilter: RECEIVED_TOPIC, qos: mqtt5_packet.QoS.AtLeastOnce}
+                { topicFilter: RECEIVED_TOPIC, qos: mqtt5.QoS.AtLeastOnce }
             ]
         });
 
-        context.subscriptions.push(topicFilter);
+        context.subscriptions[index].push(topicFilter);
         context.mqttStats.subscribesSucceeded++;
     } catch (err) {
         context.mqttStats.subscribesFailed++;
-        context.subscriptions.filter(entry => entry !== topicFilter);
+        if(context.subscriptions[index].length > 0 )
+            context.subscriptions[index].filter(entry => entry !== topicFilter);
```

Comment:
It seems `filter` method doesn't change array, instead it creates a new array with entries that passed the filter:

```
let subscriptions = ["a", "b", "c"];
let filtered_subscriptions = subscriptions.filter(entry => entry !== "b"); 
console.log("subscriptions: " + subscriptions);
console.log("filtered_subscriptions: " + filtered_subscriptions);
```

```
subscriptions: a,b,c
filtered_subscriptions: a,c
```

But `context.subscriptions[index].push(topicFilter);` is at the very end, so filtering a new topic out is probably not needed.